python/lib/dicom/summary_make.py

```diff
@@ -0,0 +1,218 @@
+from functools import cmp_to_key
+import os
+import pydicom
+import pydicom.errors
+from lib.dicom.summary_type import Summary, Info, Patient, Scanner, Acquisition, DicomFile, OtherFile
+from lib.dicom.text import make_hash, read_dicom_date_none
+
+
+def get_value(dicom: pydicom.Dataset, tag: str):
+    """
+    Get a required value from a DICOM.
+    """
+    if tag not in dicom:
+        raise Exception(f'Expected DICOM tag \'{tag}\' but found none.')
+
+    return dicom[tag].value
+
+
+def get_value_none(dicom: pydicom.Dataset, tag: str):
+    """
+    Get a nullable value from a DICOM.
+    """
+    if tag not in dicom:
+        return None
+
+    return dicom[tag].value or None
+
+
+def cmp_int_none(a: int | None, b: int | None):
+    """
+    Order comparison between two nullable integers.
+    """
+    match a, b:
+        case None, None:
+            return 0
+        case _, None:
+            return -1
+        case None, _:
+            return 1
+        case a, b:
+            return a - b
+
+
+def cmp_string_none(a: str | None, b: str | None):
+    """
+    Order comparison between two nullable strings.
+    """
+    match a, b:
+        case None, None:
+            return 0
+        case _, None:
+            return -1
+        case None, _:
+            return 1
+        case a, b if a < b:
+            return -1
+        case a, b if a > b:
+            return 1
+        case a, b:
+            return 0
+
+
+def cmp_files(a: DicomFile, b: DicomFile):
+    """
+    Compare the order of two files to sort them in the summary.
+    """
+    return \
+        cmp_int_none(a.series_number, b.series_number) or \
+        cmp_int_none(a.file_number, b.file_number) or \
+        cmp_int_none(a.echo_number, b.echo_number)
+
+
+def cmp_acquis(a: Acquisition, b: Acquisition):
+    """
+    Compare the order of two acquisitions to sort them in the summary.
+    """
+    return \
+        a.series_number - b.series_number or \
+        cmp_string_none(a.sequence_name, b.sequence_name)
+
+
+def get_dir_files(prefix: str, path: str) -> list[str]:
+    """
+    Recursively get the files of a directory.
+    """
+    if os.path.isdir(prefix + '/' + path):
+        files = []
+        for file in os.listdir(prefix + '/' + path):
+            files += get_dir_files(prefix, path + '/' + file)
+
+        # Flatten the lists of files
+        return files
+
+    return [path]
```

Comment:
This function could be moved to `lib.utilities` and could be used by other processes since it is not specific to DICOMs.

Reply:
Good point, done !